master/internal/core_checkpoint.go

```diff
@@ -85,6 +85,31 @@ func (m *Master) getCheckpointStorageConfig(id uuid.UUID) (
 	return ptrs.Ptr(legacyConfig.CheckpointStorage()), nil
 }
 
+func (m *Master) checkCheckpointAccess(
+	ctx context.Context, id uuid.UUID, content io.Writer,
+) error {
+	// Assume a checkpoint always has experiment configs
+	storageConfig, err := m.getCheckpointStorageConfig(id)
+	switch {
+	case err != nil:
+		return echo.NewHTTPError(http.StatusInternalServerError,
+			fmt.Sprintf("unable to retrieve experiment config for checkpoint %s: %s",
+				id.String(), err.Error()))
+	case storageConfig == nil:
+		return echo.NewHTTPError(http.StatusNotFound,
+			fmt.Sprintf("checkpoint not found: %s", id.String()))
+	}
+
+	_, err = checkpoints.NewDownloader(content, id.String(), storageConfig,
```

Comment:
if you are not planning on writing checkpoint content to the response, why are you passing `content` in here?

I would argue that accepting a `content io.Writer` as a parameter here can only really cause bugs, since this check function should never be writing actual byte content anywhere.

Reply:
Additionally, just calling `checkpoints.NewDownloader()` isn't actually going to tell you if you have checkpoint access.  It'll tell you if you are using S3 or not, but you actually have to read from the checkpoint a bit to find out if you have bucket listing permissions and object read permissions (which are both needed for checkpoint download).

Since it requires S3 access, which costs money, you'll also want to take steps to either:
- make sure you are not calling this too often (only when the user opens a download menu or something, like they actually expect to download), or
- configure some sort of per-bucket access cache, so the master doesn't actually have to do a mini-download to know if it has checkpoint storage access for a given bucket.